### PR TITLE
Update release configuration -use flit for pulishing

### LINKED ISSUE #412: update repository configuration so version update is not required in toml
align with kiota repositories to reduce the maintenance burden. https://github.com/microsoft/kiota-abstractions-python/pull/278#issuecomment-2124645609

### PR DESCRIPTION
**Pull request by shemogumbe**
Closes https://github.com/microsoftgraph/msgraph-beta-sdk-python/issues/412

-Update publishing tool from python build to flit.

### CODE REVIEW

**Review by andrueastman — COMMENTED** [review]
Does this mean we can drop the version entry in the ``toml`` file as well?

**Comment by shemogumbe** [comment]
> Does this mean we can drop the version entry in the `toml` file as well?

In my view, Nope - the workflow is triggered when there is a new change, with a tag starting with v, we still need to ensure the toml entry for version matches the pushed tag

**Comment by andrueastman** [comment]
I ask because the issue was created because the abstraction library does not have a version in the toml file. 

https://github.com/microsoft/kiota-abstractions-python/blob/main/pyproject.toml

So, the change was to make it easier by only having to update the version in one place(which is at https://github.com/microsoftgraph/msgraph-sdk-python/blob/main/msgraph/_version.py )

**Comment by shemogumbe** [comment]
> I ask because the issue was created because the abstraction library does not have a version in the toml file.
> 
> https://github.com/microsoft/kiota-abstractions-python/blob/main/pyproject.toml
> 
> So, the change was to make it easier by only having to update the version in one place(which is at https://github.com/microsoftgraph/msgraph-sdk-python/blob/main/msgraph/_version.py )

Made changes to update version in toml based on the tag

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -20,30 +51,28 @@ jobs:
     needs: [build]
     steps:
       - name: Checkout code
-        uses: actions/checkout@v4
+        uses: actions/checkout@v2
```
Comment (andrueastman) [inline .github/workflows/publish.yml]: Is the version downgrade intentional? (Same for line 56,58 and 75)
```suggestion
        uses: actions/checkout@v4
```

**Review by andrueastman — DISMISSED** [review]

**Review by samwelkanda — COMMENTED** [review]
Since this issue was created with the goal of standardization, do we also plan to update the core libraries to use this format as well?

**Comment by andrueastman** [comment]
I believe so, separate issue for core lib exists at https://github.com/microsoftgraph/msgraph-sdk-python-core/issues/573 and is already assigned to @shemogumbe

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -24,18 +55,16 @@ jobs:
       - name: Set up Python
         uses: actions/setup-python@v5
         with:
-          python-version: '3.12'
-      - name: Install dependencies
+          python-version: 3.9
```
Comment (baywet) [inline .github/workflows/publish.yml]: Why the downgrade?

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -9,7 +9,38 @@ permissions:
...
+    name: Update version
+    runs-on: ubuntu-latest
+    if: startsWith(github.ref, 'refs/tags/v')
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: 3.9
```
Comment (baywet) [inline .github/workflows/publish.yml]: Why not 12?

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -24,18 +55,16 @@ jobs:
...
-      - name: Publish package
-        uses: pypa/gh-action-pypi-publish@81e9d935c883d0b210363ab89cf05f3894778450
-        with:
-          user: __token__
-          password: ${{ secrets.PYPI_API_TOKEN }}
+          pip install flit
+      - name: Publish the distibution to PyPI
+        run: flit publish
+        env:
+          FLIT_INDEX_URL: https://upload.pypi.org/legacy/
```
Comment (baywet) [inline .github/workflows/publish.yml:61]: For my education, are we using the legacy endpoint because we're not using OIDC?

**Review by baywet — CHANGES_REQUESTED** [review]

File: .github/workflows/publish.yml (hunk 4 above)
Comment (shemogumbe) [reply .github/workflows/publish.yml:61]: The term legacy here does not mean this is deprecated or outdated, the legacy endpoint is recommended for uploads with twine or flit while the 'new' endpoint is recommended for web uploads https://warehouse.pypa.io/api-reference/legacy.html#upload-api

**Comment by baywet** [comment]
getting there, I made my comments on the beta PR for some reason this time around...

**Comment by baywet** [comment]
again @shemogumbe please see the latest recommended changes there https://github.com/microsoftgraph/msgraph-beta-sdk-python/pull/424, and once the updates are done in beta, replicate here.

**Review by baywet — APPROVED** [review]

**Pull request merged**